Steplist component

dgrossen opened the pull request:
Reviving work on the StepList component and dusting off PR https://github.com/adobe/react-spectrum/pull/2832

- Fixed some bugs with rtl and the use of solid/dashed lines for vertical completed steps
- Updated stories
- Cleaned up dependencies

Still need to add tests and docs but thought I would get the ball rolling.

StepList Component Issue - https://github.com/adobe/react-spectrum/issues/1012

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

<!--- Include instructions to test this pull request -->

## 🧢 Your Project:

<!--- Company/project for pull request -->

snowystinger commented on packages/@react-spectrum/steplist/test/StepList.test.js:
```diff
@@ -0,0 +1,29 @@
+import StepList from '../';
+import {render} from '@testing-library/react';
+import React from 'react';
+import V2StepList from '@react/react-spectrum/StepList';
```
you can delete this import

snowystinger commented on packages/@react-spectrum/steplist/test/StepList.test.js:
```diff
@@ -0,0 +1,29 @@
+  it.each`
+    Name | Component      | props
+    ${'StepList'} | ${StepList}| ${{}}
+    ${'V2StepList'}      | ${V2StepList}      | ${{}}
```
and this line, we're not testing against v2 anymore

snowystinger commented on packages/@react-stately/steplist/src/useStepListState.ts:
```diff
@@ -0,0 +1,88 @@
+
+  let [lastCompletedStep, setLastCompletedStep] = useControlledState<Key>(props.lastCompletedStep, props.defaultLastCompletedStep, props.onLastCompletedStepChange);
+  const {setSelectedKey: realSetSelectedKey, selectedKey, collection} = state;
+  const {prevKeyMap, indexMap} = useMemo(() => buildKeysMaps(collection), [collection]);
```
is this just for keeping track of the currently selected key and then marking all before it as completed?
instead of building this map, could we just use collection.getKeyBefore?
could just keep a Set of all keys before the selected one, might be a little clearer?
instead of `indexMap.get(step) <= indexMap.get(lastCompletedStep);` you could just write `completed.has(key)`

snowystinger replied:
also, i'm not quite following why it's called prevKeyMap, it's current?

ooo, I just figured out this means, previous key... map, not previous... key map. if we're going to keep it, then lets rename it. mapOfPreviousKeys? keysLinkedList?

snowystinger commented on packages/@react-stately/steplist/src/useStepListState.ts:
```diff
@@ -0,0 +1,88 @@
+  return {
+    ...state,
+    setSelectedKey,
+    setLastCompletedStep,
```
where is this used? can it be controlled through props `selectedKey`?

snowystinger reviewed (commented):
I don't seem able to undo any step except the last one
steps:
default story, keyboard navigate to the steplist and then to the last item
keyboard navigate back to the first item, only the last step is cleared out

Rest of my comments I think could be candidates for followup:

would be good to have less stories if they are covered by the controls, and instead add the more edge case stories, such as ones with steps that have long names which wrap or cause the steplist to take 100% of the width, or show them in a dialog, or maybe one with controlled selection that updates in response to events, or a disabled key in the middle but still have some enabled toward the far end. or starting with a disabled key. it'd also be good to have one with the keyboard activation set to manual

I have a feeling that we'll want/need to support links in steplists, we'll need an example of that to make sure it works
what I mean by that is specifically https://react-spectrum.adobe.com/react-spectrum/Breadcrumbs.html#links

dgrossen commented:
@snowystinger please give it another review when you get a chance. I addressed most of the feedback here and several of the items from the testing bash. I also cleaned up some things. Opted to use context instead of passing the `state` as a prop to the items. I also got rid of the `isNavigable` stuff to align more with the #1012 spec. It should also follow the XD spec more closely also

snowystinger commented on packages/@react-aria/steplist/test/useStepList.test.js:
```diff
@@ -0,0 +1,17 @@
+ * governing permissions and limitations under the License.
+ */
+
+describe('useStepList', function () {
```
can just delete this file

snowystinger commented on packages/@react-spectrum/steplist/chromatic/StepList.chromatic.tsx:
```diff
@@ -0,0 +1,40 @@
+
+export default meta;
+
+const Template = (args: SpectrumStepListProps<object>) => (
```
if any more are added to this, we should update it to storybook CSF 3.0 format

snowystinger commented on packages/@react-spectrum/steplist/src/StepList.tsx:
```diff
@@ -0,0 +1,72 @@
+  let {listProps} = useStepList(props, state, domRef);
+
+  const {isDisabled, isEmphasized, isReadOnly} = props;
+  const stepListItems = [...state.collection].map((item) => (
```
easier to read if this is down in the render, nothing to be gained having out here, no conditional wrappers or anything

snowystinger commented on packages/@react-spectrum/steplist/stories/StepList.stories.tsx:
```diff
@@ -0,0 +1,205 @@
+export type StepListStory = ComponentStoryObj<typeof StepList>;
+export type AsyncItemsStory = ComponentStoryObj<typeof AsyncStepListItems>;
+
+export const Default: DefaultStory = {
```
can delete all the other stories (except disabled keys), they're all captured by controls. we'll still need those edge case ones I mentioned in testing, I'm happy to help if you're low on time

snowystinger commented on packages/@react-spectrum/steplist/test/StepList.test.js:
```diff
@@ -0,0 +1,30 @@
+import {render} from '@react-spectrum/test-utils';
+import {StepList} from '../';
+
+describe('StepList', function () {
```
we need tests for this component, happy to help

snowystinger commented:
Some behavioral things I noticed:
Up and Down arrow keys will still navigate
Can't click on any of the steps anymore really in any of the stories because there are so many disabled?
I don't know if most of the stories are missing an action for 'onAction', but I don't see onAction being fired in any of the stories
What is the Default story? it doesn't seem to do anything? I think it got mixed with the fully disabled story?

In the last story, when I press next, I can go to step 4, but then if I click on step < 4, I can't get back to step 4 without clicking the next button eventually

Unsure about this one, but when individual steps are disabled, they don't display their number, yet steps around them will. I think it looks funky, not sure I like it.

dgrossen commented:
> Up and Down arrow keys will still navigate 

Fixed
> Can't click on any of the steps anymore really in any of the stories because there are so many disabled?

This is because I made changes to align more with the spec: https://github.com/adobe/react-spectrum/issues/1012
Specifically, "Users can optionally click on completed steps to jump back to them, but advancing to the next step must be done through alternate means" and "By default, the last completed step advances automatically as the selected step increases". And I spoke with XD and they confirmed that expected behavior. Sorry for the confusion, the first story/implementation allowed any steps to be clicked on.

> In the last story, when I press next, I can go to step 4, but then if I click on step < 4, I can't get back to step 4 without clicking the next button eventually

I improved the stories a bit to help clear things up. I tried to rename them and added a `last completed step` display to help communicate what's going on. The back button and next button are disabled when on the first and last step respectively.

I also added tests that hopefully describe the expected controlled and uncontrolled (sort of) states of the step list component.

> Unsure about this one, but when individual steps are disabled, they don't display their number, yet steps around them will. I think it looks funky, not sure I like it.

I wasn't able to repro this one, but I wondered if this was related to the designs not having the numbers in incomplete steps for medium size, but the large and extra large incomplete steps do? https://spectrum-contributions.corp.adobe.com/page/steplist-beta/

snowystinger reviewed (commented):
I cannot tab to the steplist in the Default story, I think the current step should be part of navigation at the very least? Otherwise it's hard to see that you're starting a multi step process as a keyboard user with VO. I think it also makes sense to be able to look ahead at future steps.

In the second one, 'Default - completed', I can tab to it, but when I press 'Enter' on 1 or 2, and then continue tabbing through, it leaves focus rings behind
<img width="522" alt="Screenshot 2023-10-19 at 6 55 49 AM" src="https://github.com/adobe/react-spectrum/assets/698229/b918cc22-043e-47c5-8e59-eee3ae9bc03d">

General, there is no visual difference between current step and the yet to be visited future steps, discovery that I can click back on ones that have lost their number might be hard.


I like the new stories, helps a lot with getting a feel for this.

dgrossen commented on packages/@react-spectrum/steplist/test/StepList.test.js:
```diff
@@ -0,0 +1,273 @@
+    expect(stepListItems[1]).toHaveClass('is-completed');
+  });
+
+  it('does not updated selected step when last completed step is controlled', function () {
```
It was a typo. Good catch

dgrossen commented on packages/@adobe/spectrum-css-temp/components/steplist/index.css:
```diff
@@ -13,255 +13,307 @@ governing permissions and limitations under the License.
+      line-height: var(--spectrum-steplist-marker-diameter-small);
+      width: var(--spectrum-steplist-marker-diameter-small);
+      height: var(--spectrum-steplist-marker-diameter-small);
+      font-size: 0;
```
Looks like it hides the numbers when the size is small

snowystinger replied:
ah, tricky

dgrossen commented on packages/@react-spectrum/steplist/test/StepList.test.tsx:
```diff
@@ -214,8 +217,8 @@ describe('StepList', function () {
 
-    expect(onLastCompletedStepChange).toHaveBeenCalledTimes(0);
-    expect(stepListItems[0].firstElementChild.textContent).toContain('Completed');
+    expect(onLastCompletedStepChange).toHaveBeenCalledWith('step-one');
```
Hey @snowystinger thanks for contributing! You've added an assertion here to check `onLastCompletedStepChange` has been fired with `step-one` but the component was given a `defaultLastCompletedStep="step-one"` prop and the prop hasn't changed. It should only change once step-two has completed and step-two only completes once the selected key has advanced to step-three.

dgrossen replied:
After reading https://github.com/adobe/react-spectrum/issues/5013#issuecomment-1703101568

Are you suggesting it should be called after the initial render? Similar to `onSelectionChange`?

snowystinger replied:
sorry, missed the default or it my mind break, you are correct, it should not be called until after we've advanced it to step two

dgrossen commented on packages/@react-spectrum/steplist/test/StepList.test.tsx:
```diff
@@ -228,17 +231,18 @@ describe('StepList', function () {
         id="steplist-id"
         aria-label="steplist-test"
         lastCompletedStep={'step-one'}
-        onSelectedChange={onSelectedChange}>
```
oh my, sorry I misspelled this again!

snowystinger commented on packages/@react-spectrum/steplist/package.json:
```diff
@@ -0,0 +1,68 @@
+    "@react-stately/collections": "^3.10.2",
+    "@react-stately/list": "^3.10.0",
+    "@react-stately/steplist": "^3.0.0-alpha.1",
+    "@react-stately/tree": "^3.7.3",
```
unused, can remove

snowystinger commented on packages/@react-spectrum/steplist/package.json:
```diff
@@ -0,0 +1,68 @@
+    "@react-spectrum/layout": "^3.5.7",
+    "@react-spectrum/utils": "^3.10.2",
+    "@react-stately/collections": "^3.10.2",
+    "@react-stately/list": "^3.10.0",
```
not used directly in this package, can remove

snowystinger commented on packages/@react-spectrum/steplist/package.json:
```diff
@@ -0,0 +1,68 @@
+    "@react-aria/steplist": "^3.0.0-alpha.1",
+    "@react-aria/utils": "^3.20.0",
+    "@react-aria/visually-hidden": "^3.8.4",
+    "@react-spectrum/layout": "^3.5.7",
```
only used in storybook, can remove

snowystinger reviewed (commented):
we should check all package jsons so we don't have extraneous dependencies, I did a spot check

LFDanLu commented on packages/@react-aria/steplist/docs/useStepList.mdx:
fine for now, note to self to add a followup ticket to flesh out the docs. We might want to add a `after_version` or something to prevent this page from going out accidentally in a release or get rid of it entirely

LFDanLu commented on packages/@adobe/spectrum-css-temp/components/steplist/skin.css:
For followup: address HCM styling, especially for disabled steplist

LFDanLu commented on packages/@react-aria/steplist/src/useStepList.ts:
```diff
@@ -0,0 +1,51 @@
+
+  const strings = useLocalizedStringFormatter(intlMessages);
+  const stepListProps: HTMLAttributes<HTMLElement> = {
+    ...mergeProps(listProps, filterDOMProps(props)),
```
Should this be `filterDOMProps(props, {labelable: true})` so that things like `aria-labelledby` get processed automatically?

LFDanLu commented on packages/@react-spectrum/steplist/docs/StepList.mdx:
might need a `after_version` as well so it doesn't get pushed to prod docs on our release

LFDanLu commented on packages/@react-spectrum/steplist/intl/index.js:
don't think we need this file

LFDanLu commented on packages/@react-spectrum/steplist/intl/en-US.json:
```diff
@@ -0,0 +1,6 @@
+  "completed": "Completed: ",
+  "notCompleted": "Not completed: ",
+  "current": "Current: ",
+  "steplist": "Step List"
```
don't think we need this line since the aria hook the the one that needs this translation?

dgrossen replied:
This is for localizing the VisuallyHidden component states of the step to the DOM:

```
<li>
  <a role="link" tabIndex="0">
    <VisuallyHidden>Completed: </VisuallyHidden>
    Details
  </a>
</li>
```

LFDanLu replied:
Right, but we don't need `"steplist"` if I'm not mistaken? Only the previous three lines are used: https://github.com/adobe/react-spectrum/blob/060af1459b91d76a15d6d165f0d6b6902e222c4f/packages/%40react-spectrum/steplist/src/StepListItem.tsx#L55-L61

reidbarber reviewed (commented):
I think the isEmphasized text color may fail contrast requirements, so we may want to revisit that design. Besides that, I think this looks ready for an alpha release (after the empty documentation pages get removed.)

snowystinger reviewed (approved):
Tracking remaining work for the team here https://github.com/orgs/adobe/projects/19/views/4?pane=issue&itemId=42835801

LFDanLu reviewed (approved):
Happy for this to be merged as is for testing and for the rest to be followup
